cpu/stm32f2/periph/spi.c

```diff
@@ -16,440 +18,143 @@
-
-    while (!(spi_port->SR & SPI_SR_RXNE));
-
-    if (in != NULL) {
-        *in = spi_port->DR;
+    /* lock bus */
+    mutex_lock(&locks[bus]);
+    /* enable SPI device clock */
+    periph_clk_en(spi_config[bus].abpbus, spi_config[bus].rccmask);
+    /* enable device */
+    dev(bus)->CR1 = (((clk + spi_config[bus].abpbus) << 3) | mode | SPI_CR1_MSTR);
+    dev(bus)->CR2 = 0;
+    if (cs != SPI_HWCS_MASK) {
+        dev(bus)->CR1 |= (SPI_CR1_SSM | SPI_CR1_SSI);
     }
     else {
-        spi_port->DR;
+        dev(bus)->CR2 |= (SPI_CR2_SSOE);
     }
 
-    return 1;
+    return SPI_OK;
 }
 
-void spi_transmission_begin(spi_t dev, char reset_val)
+void spi_release(spi_t bus)
 {
-
-    switch (dev) {
-#if SPI_0_EN
-        case SPI_0:
-            SPI_0_DEV->DR = reset_val;
-            break;
-#endif
-#if SPI_1_EN
-        case SPI_1:
-            SPI_1_DEV->DR = reset_val;
-            break;
-#endif
-#if SPI_2_EN
-        case SPI_2:
-            SPI_2_DEV->DR = reset_val;
-            break;
-#endif
-    }
+    /* disable device and release lock */
+    dev(bus)->CR1 = 0;
+    periph_clk_dis(spi_config[bus].abpbus, spi_config[bus].rccmask);
+    mutex_unlock(&locks[bus]);
 }
 
-void spi_poweron(spi_t dev)
+void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
+                        const void *out, void *in, size_t len)
 {
-    switch (dev) {
-#if SPI_0_EN
-        case SPI_0:
-            SPI_0_CLKEN();
-            break;
-#endif
-#if SPI_1_EN
-        case SPI_1:
-            SPI_1_CLKEN();
-            break;
-#endif
-#if SPI_2_EN
-        case SPI_2:
-            SPI_2_CLKEN();
-            break;
-#endif
-    }
-}
+    uint8_t *inbuf = (uint8_t *)in;
+    uint8_t *outbuf = (uint8_t *)out;
 
-void spi_poweroff(spi_t dev)
-{
-    switch (dev) {
-#if SPI_0_EN
-        case SPI_0:
-            while (SPI_0_DEV->SR & SPI_SR_BSY);
-            SPI_0_CLKDIS();
-            break;
-#endif
-#if SPI_1_EN
-        case SPI_1:
-            while (SPI_1_DEV->SR & SPI_SR_BSY);
-            SPI_1_CLKDIS();
-            break;
-#endif
-#if SPI_2_EN
-        case SPI_2:
-            while (SPI_2_DEV->SR & SPI_SR_BSY);
-            SPI_2_CLKDIS();
-            break;
-#endif
-    }
-}
+    /* make sure at least one input or one output buffer is given */
+    assert(outbuf || inbuf);
 
-static inline void irq_handler_transfer(SPI_TypeDef *spi, spi_t dev)
-{
+    /* active the given chip select line */
+    dev(bus)->CR1 |= (SPI_CR1_SPE);     /* this pulls the HW CS line low */
+    if ((cs != SPI_HWCS_MASK) && (cs != SPI_CS_UNDEF)) {
+        gpio_clear((gpio_t)cs);
+    }
 
-    if (spi->SR & SPI_SR_RXNE) {
-        char data;
-        data = spi->DR;
-        data = spi_config[dev].cb(data);
-        spi->DR = data;
+    /* transfer data, use shortpath if only sending data */
+    if (!inbuf) {
+        for (size_t i = 0; i < len; i++) {
+            while (!(dev(bus)->SR & SPI_SR_TXE));
+            dev(bus)->DR = outbuf[i];
+        }
```

Comment:
Sorry, just found another issue. Even if you don't read, you should empty the buffer. Add these lines:
```
            while (!(dev(bus)->SR & SPI_SR_RXNE)) {}
            dev(bus)->DR; // read
```